hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/DefaultS3ClientFactory.java

```diff
@@ -94,9 +105,81 @@ public AmazonS3 createS3Client(
       awsConf.setUserAgentSuffix(parameters.getUserAgentSuffix());
     }
 
-    return buildAmazonS3Client(
-        awsConf,
-        parameters);
+    if (conf.get(CLIENT_SIDE_ENCRYPTION_METHOD) == null) {
+      return buildAmazonS3Client(
+          awsConf,
+          parameters);
+    } else {
+      return newAmazonS3EncryptionClient(
+          awsConf,
+          parameters);
+    }
+  }
+
+  /**
+   * Create an {@link AmazonS3} client of type
+   * {@link AmazonS3EncryptionV2} if CSE is enabled.
+   *
+   * @param awsConf    AWS configuration.
+   * @param parameters parameters
+   *
+   * @return new AmazonS3 client.
+   */
+  protected AmazonS3 newAmazonS3EncryptionClient(
+      final ClientConfiguration awsConf,
+      final S3ClientCreationParameters parameters){
+
+    AmazonS3 client;
+    AmazonS3EncryptionClientV2Builder builder =
```

Comment:
As this extends AmazonS3Builder, it should be possible for all the common setup code (endpoints, regions etc) to be pulled out into a shared method. This will help us the next time we have to do emergency fixes, and will ensure that changes to the normal path also get picked up by by the new client